File: iOS/Inkwell/Authentication/InkwellOAuthScopes.swift
//
//  InkwellOAuthScopes.swift
//  Inkwell
//
//  Construction of the OAuth scope list Inkwell requests at authorization time.
//

import Foundation

/// Inkwell's OAuth scope request, and the rules that keep it aligned with the hosted
/// `client-metadata.json`.
///
/// This mirrors `InkwellOAuthScopes` / `StandardSitePermissionSets` in shared KMP
/// (`shared/src/commonMain/kotlin/uk/ewancroft/inkwell/shared/oauth/`). Kotlin is the
/// canonical definition; this file exists because the checked-in
/// `InkwellShared.xcframework` predates those types, exactly like the hardcoded
/// `sharedOAuthScopeRepoLeafletComment()` in `SharedKMP.swift`. Rebuild the framework
/// and this file collapses into a thin wrapper.
///
/// ## The superset invariant
///
/// An Authorization Server checks each requested scope by plain string membership
/// against the `scope` field of the client metadata document — it does not expand
/// permission sets before that check. So ``requested`` must be a literal string subset
/// of what `client-metadata.json` declares, and adding a scope here without publishing
/// it there breaks every login.
enum InkwellOAuthScopes {

    /// Whether the Standard.site block is requested as `include:site.standard.authFull`
    /// rather than as the four granular `repo:site.standard.*` scopes it expands to.
    ///
    /// **Mirror of `StandardSitePermissionSets.USE_PERMISSION_SET` in shared KMP — keep
    /// the two in sync.**
    ///
    /// Kept `false` until a human has confirmed a real authorization flow against both a
    /// Bluesky-hosted PDS and a self-hosted one. The published `site.standard.authFull`
    /// lexicon grants a `repo` permission over the same four collections with no
    /// `action` restriction, which is exactly what the granular scopes already grant, so
    /// flipping this changes the consent screen wording and nothing about the effective
    /// permissions — no existing user needs to re-authorize. Hosted client metadata
    /// already declares both forms, so no metadata redeploy is needed either.
    ///
    /// The reason it is not already `true`: requesting a permission set makes login
    /// depend on the user's PDS being able to *resolve* the `site.standard.authFull`
    /// lexicon (DNS → PLC → publishing PDS) at session start, and neither OAuthenticator
    /// nor the Android OAuth library exposes a seam to retry a rejected pushed
    /// authorization request with a different scope string. There is no field in
    /// `oauth-authorization-server` metadata that advertises permission-set support, so
    /// there is no honest pre-flight capability check either.
    static let usePermissionSet = false

    /// `include:site.standard.authFull` — the permission set covering publication,
    /// document, subscription and recommend records.
    static var includeAuthFull: String { "include:" + sharedOAuthScopeAuthFull() }

    /// `include:site.standard.authSocial` — subscriptions and recommends only.
    ///
    /// Unused: Inkwell has one sign-in path and it always leads to the Writer tab, so a
    /// session that cannot create publication or document records would fail on first
    /// publish. This becomes the right request only if a genuinely read-only sign-in
    /// mode is added.
    static var includeAuthSocial: String { "include:" + sharedOAuthScopeAuthSocial() }

    /// The granular scopes `site.standard.authFull` expands to, in the order the
    /// published lexicon lists its collections.
    static var standardSiteGranular: [String] {
        [
            sharedOAuthScopeRepoPublication(),
            sharedOAuthScopeRepoDocument(),
            sharedOAuthScopeRepoSubscription(),
            sharedOAuthScopeRepoRecommend(),
        ]
    }

    /// Scopes that can never live in a permission set.
    ///
    /// The permissions spec bars `blob` permissions from permission sets outright.
    static var alwaysExplicit: [String] {
        [sharedOAuthScopeAtproto(), sharedOAuthScopeBlobAll()]
    }

    /// Inkwell-specific and third-party record scopes.
    ///
    /// None sit under the `site.standard` namespace, so no Standard.site permission set
    /// is permitted to grant them — a set may only address resources in its own NSID
    /// hierarchy.
    static var appRecordScopes: [String] {
        [
            sharedOAuthScopeRepoLeafletComment(),
            sharedOAuthScopeRepoUserInputDiscussion(),
            "repo:uk.ewancroft.inkwell.user",
        ]
    }

    /// Personal moderation: block records plus the Bluesky AppView RPCs behind the
    /// mute/block/report UI.
    static let moderationScopes = [
        "repo:app.bsky.graph.block?action=create&action=delete",
        "rpc:app.bsky.graph.muteActor?aud=did:web:api.bsky.app%23bsky_appview",
        "rpc:app.bsky.graph.unmuteActor?aud=did:web:api.bsky.app%23bsky_appview",
        "rpc:app.bsky.graph.getMutes?aud=did:web:api.bsky.app%23bsky_appview",
        "rpc:app.bsky.graph.getBlocks?aud=did:web:api.bsky.app%23bsky_appview",
        "rpc:com.atproto.moderation.createReport?aud=did:web:api.bsky.app%23bsky_appview",
    ]

    /// What the app actually asks for at authorization time.
    static func requested(usePermissionSet: Bool = InkwellOAuthScopes.usePermissionSet) -> [String] {
        alwaysExplicit
            + (usePermissionSet ? [includeAuthFull] : standardSiteGranular)
            + appRecordScopes
            + moderationScopes
    }

    /// The maximum scope declared in the hosted `client-metadata.json`: every scope the
    /// runtime could request under either setting of ``usePermissionSet``.
    static var clientMetadataDeclared: [String] {
        alwaysExplicit
            + standardSiteGranular
            + [includeAuthFull]
            + appRecordScopes
            + moderationScopes
    }

    /// Expands a known `include:` scope into the granular scopes it grants, or `nil` if
    /// it is not a permission set Inkwell models.
    ///
    /// Only the two Standard.site sets are modelled; guessing at a third party's set
    /// would be worse than admitting ignorance.
    static func expand(_ scope: String) -> [String]? {
        switch scope {
        case includeAuthFull: return standardSiteGranular
        case includeAuthSocial:
            return [sharedOAuthScopeRepoSubscription(), sharedOAuthScopeRepoRecommend()]
        default: return nil
        }
    }

    /// The effective granular permissions a scope list resolves to, with every known
    /// permission set expanded in place.
    ///
    /// Used to show that flipping ``usePermissionSet`` grants the same thing.
    static func effectivePermissions(_ scopes: [String]) -> Set<String> {
        Set(scopes.flatMap { expand($0) ?? [$0] })
    }
}

File: shared/src/commonMain/kotlin/uk/ewancroft/inkwell/shared/oauth/InkwellOAuthScopes.kt
package uk.ewancroft.inkwell.shared.oauth

/**
 * The single source of truth for Inkwell's OAuth scope request and for the maximum
 * scope its hosted `client-metadata.json` declares.
 *
 * ## The superset invariant
 *
 * The reference Authorization Server checks a requested scope by plain string
 * membership against the `scope` field of the client metadata document
 * (`oauth-provider/src/client/client.ts`: `Scope "<x>" is not declared in the client
 * metadata`). It does **not** expand permission sets before that check, and it does not
 * parse individual metadata scope tokens at registration time beyond requiring
 * `atproto` and rejecting duplicates
 * (`oauth-provider/src/client/client-manager.ts`).
 *
 * Two consequences drive the design here:
 *
 * 1. [runtimeScopes] must be a literal string subset of [clientMetadataScopes]. Not an
 *    "effectively equivalent" subset — the same strings.
 * 2. Declaring `include:site.standard.authFull` in client metadata is inert on an
 *    Authorization Server that does not understand permission sets, because nothing
 *    validates metadata scope tokens. So metadata can safely declare both the
 *    permission set and the granular scopes it expands to, and the runtime can pick
 *    either without a metadata redeploy.
 *
 * The three hosted copies of the metadata document must all carry
 * [clientMetadataScopeString]:
 *  - `website/src/routes/client-metadata.json/+server.ts` (the live one PDSes fetch)
 *  - `iOS/oauth/client-metadata.json`
 *  - `Android/docs/oauth/client-metadata.json`
 */
object InkwellOAuthScopes {

    /**
     * Scopes that cannot live in a permission set and must always be requested
     * directly.
     *
     * The permissions spec bars `blob` permissions from permission sets outright:
     * "Permissions of this type can not be included in permission sets, and must be
     * requested directly by client apps."
     */
    val ALWAYS_EXPLICIT: List<String> = listOf(
        OAuthScopes.ATPROTO,
        OAuthScopes.BLOB_ALL,
    )

    /**
     * Inkwell-specific and third-party record scopes.
     *
     * None of these sit under the `site.standard` namespace, so no Standard.site
     * permission set is allowed to grant them — permission sets may only address
     * resources in their own NSID hierarchy.
     */
    val APP_RECORD_SCOPES: List<String> = listOf(
        OAuthScopes.REPO_LEAFLET_COMMENT,
        OAuthScopes.REPO_USERINPUT_DISCUSSION,
        OAuthScopes.REPO_USER,
    )

    /**
     * Personal moderation: block records plus the Bluesky AppView RPCs behind the
     * mute/block/report UI on both platforms.
     */
    val MODERATION_SCOPES: List<String> = listOf(
        "repo:app.bsky.graph.block?action=create&action=delete",
        "rpc:app.bsky.graph.muteActor?aud=did:web:api.bsky.app%23bsky_appview",
        "rpc:app.bsky.graph.unmuteActor?aud=did:web:api.bsky.app%23bsky_appview",
        "rpc:app.bsky.graph.getMutes?aud=did:web:api.bsky.app%23bsky_appview",
        "rpc:app.bsky.graph.getBlocks?aud=did:web:api.bsky.app%23bsky_appview",
        "rpc:com.atproto.moderation.createReport?aud=did:web:api.bsky.app%23bsky_appview",
    )

    /**
     * What the app actually asks for at authorization time.
     *
     * @param usePermissionSet when `true`, the four `repo:site.standard.*` scopes are
     *   replaced by the single `include:site.standard.authFull` invocation. The
     *   effective grant is identical either way — see
     *   [StandardSitePermissionSets.AUTH_FULL_EXPANSION].
     */
    fun runtimeScopes(
        usePermissionSet: Boolean = StandardSitePermissionSets.USE_PERMISSION_SET,
    ): List<String> =
        ALWAYS_EXPLICIT +
            StandardSitePermissionSets.standardSiteScopes(usePermissionSet) +
            APP_RECORD_SCOPES +
            MODERATION_SCOPES

    /**
     * The maximum scope declared in hosted client metadata: every scope the runtime
     * could request under either setting of
     * [StandardSitePermissionSets.USE_PERMISSION_SET].
     *
     * Declaring more than the app currently requests costs the user nothing — the
     * consent screen shows the *requested* scopes, not the declared ones — and removes
     * the deploy ordering hazard where a client update ships before the metadata that
     * authorizes it.
     */
    fun clientMetadataScopes(): List<String> =
        ALWAYS_EXPLICIT +
            StandardSitePermissionSets.AUTH_FULL_EXPANSION +
            StandardSitePermissionSets.INCLUDE_AUTH_FULL +
            APP_RECORD_SCOPES +
            MODERATION_SCOPES

    /** Space-joined form of [runtimeScopes], as sent in the OAuth `scope` parameter. */
    fun runtimeScopeString(
        usePermissionSet: Boolean = StandardSitePermissionSets.USE_PERMISSION_SET,
    ): String = runtimeScopes(usePermissionSet).joinToString(" ")

    /** Space-joined form of [clientMetadataScopes], as published in `client-metadata.json`. */
    val clientMetadataScopeString: String
        get() = clientMetadataScopes().joinToString(" ")
}
